src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.Extensions.cs

```diff
@@ -12,13 +12,94 @@ public static unsafe partial class Vector
+        {
+            if (((nuint)destination % (uint)(Vector2.Alignment)) != 0)
+            {
+                ThrowHelper.ThrowAccessViolationException();
+            }
+
+            *(Vector2*)destination = source;
+        }
+
+        /// <summary>Stores a vector at the given 8-byte aligned destination.</summary>
+        /// <param name="source">The vector that will be stored.</param>
+        /// <param name="destination">The aligned destination at which <paramref name="source" /> will be stored.</param>
+        /// <exception cref="AccessViolationException"><paramref name="destination" /> is not 8-byte aligned.</exception>
+        /// <remarks>This method may bypass the cache on certain platforms.</remarks>
+        [CLSCompliant(false)]
+        public static void StoreAlignedNonTemporal(this Vector2 source, float* destination) => source.StoreAligned(destination);
+
+        /// <summary>Stores a vector at the given destination.</summary>
+        /// <param name="source">The vector that will be stored.</param>
+        /// <param name="destination">The destination at which <paramref name="source" /> will be stored.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void StoreUnsafe(this Vector2 source, ref float destination)
+        {
+            ref byte address = ref Unsafe.As<float, byte>(ref destination);
+            Unsafe.WriteUnaligned(ref address, source);
+        }
+
+        /// <summary>Stores a vector at the given destination.</summary>
+        /// <param name="source">The vector that will be stored.</param>
+        /// <param name="destination">The destination to which <paramref name="elementOffset" /> will be added before the vector will be stored.</param>
+        /// <param name="elementOffset">The element offset from <paramref name="destination" /> from which the vector will be stored.</param>
+        [CLSCompliant(false)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void StoreUnsafe(this Vector2 source, ref float destination, nuint elementOffset)
+        {
+            destination = ref Unsafe.Add(ref destination, (nint)elementOffset);
+            Unsafe.WriteUnaligned(ref Unsafe.As<float, byte>(ref destination), source);
+        }
+
+        /// <inheritdoc cref="ToScalar(Vector4)" />
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static float ToScalar(this Vector2 vector) => vector.AsVector128Unsafe().ToScalar();
+
+        /// <inheritdoc cref="WithElement(Vector4, int, float)" />
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector2 WithElement(this Vector2 vector, int index, float value)
+        {
+            if ((uint)index >= Vector2.ElementCount)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index);
+            }
+            return vector.AsVector128Unsafe().WithElement(index, value).AsVector2();
```

Comment:
When I read this, it looks like it's going to do a 16-byte read of a Vector128 from the location of the Vector2, replace one of the elements, then mask off the extra elements. Can't this overrun a buffer or otherwise do something bad?

Reply:
`AsVector128Unsafe()` converts the `Vector2` to a `Vector128` leaving the upper two elements undefined rather than explicitly zeroing them; it does not cause a read of 16-bytes of data. That is, it will still only read 64-bits (the 2x float fields), just doing so however is most efficient since the upper bits are irrelevant given the guard clause and subsequent downcast back to `Vector2`.